platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc

```diff
@@ -156,6 +156,13 @@ Calling the following methods on link:{javadocPath}/org/gradle/api/artifacts/Con
 - addLater(Provider)
 - addAllLater(Provider)
 
+[[deprecate_get_dependency_project]]
+==== Deprecated `ProjectDependency#getDependencyProject()`
```

Comment:
What should be used instead?

Currently, in AndroidX project we use it in two places:

- Validating if a dependency is publishing and has the right level of stability (e.g. stable library should not pull in beta dependencies) [code](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:buildSrc/private/src/main/kotlin/androidx/build/VerifyDependencyVersionsTask.kt;l=253?q=dependency.dependencyProject.extensions&ss=androidx%2Fplatform%2Fframeworks%2Fsupport)

- Building a graph of inter-project dependencies that is a map of project.path to project.path [code](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:buildSrc/private/src/main/kotlin/androidx/build/dependencyTracker/DependencyTracker.kt;l=39?q=.dependencyProject&sq=&ss=androidx%2Fplatform%2Fframeworks%2Fsupport)

Reply:
For the first bullet point, that should be backed by attributes and attribute compatibility rules.

If one project is "stable" and somebody tries to add a dependency that is "beta", and "stable" things cannot depend on "beta" things, you can model this with attributes. 

The (stable) consumer would add an attribute with that value to the configuration being resolved. The (beta) producer would add an attribute with that value to the consumable configuration acting as a variant. 

Then, you can write an attribute compatibility rule which enforces that consumers of type stable are not compatible with producers of type beta. 

This way, the dependency resolution engine would enforce the stable/beta constraint for you, without requiring additional build logic.

Reply:
The second use case is more tricky. Can you go into detail on why you need to build up such a graph? 

In general the approach you have will fail to account for substituted dependencies -- like those from included builds. It will only detect dependencies that are _declared_ as project dependencies, but will exclude dependencies that are originally declared as module dependencies but resolve to projects (like substitutions from included builds). 

For this reason, when analyzing dependency graphs like this, we recommend using the `ResolutionResult`, which accounts for and exposes both the _requested_ and _selected_  variants for any given dependency. 

For your particular use case, it seems like the build logic is using this project to project mapping to detect which projects need to be built given some changed files. Does up-to-date checking not solve this use case?